v3/mqtt.go

```diff
@@ -465,91 +577,171 @@ func (conn *mqttConn) runPacketWriter(wg *sync.WaitGroup) {
 		wg.Done()
 	}()
 
-	// Run until closed
-	for p := range conn.outPktCh {
-		err := conn.writePacket(p)
-		if err != nil {
-			logError("Unable to write out packet %s", p.Type())
-			switch p.Type() {
-			case packet.PINGREQ:
-				conn.pingRespCh <- false
-			case packet.SUBSCRIBE:
-				subPkt := p.(*packet.SubscribePacket)
-				conn.queueAckCh <- MqttQueueResult{subPkt.PacketID, err}
-			case packet.PUBLISH:
-				pubPkt := p.(*packet.PublishPacket)
-				conn.queueAckCh <- MqttQueueResult{pubPkt.PacketID, err}
-			default:
-				// We don't have a packet ID (or did not expect to receive this
-				// packet type and fail in sending it
-				err := fmt.Errorf("Failed to queue packet %s", p.Type())
-				conn.queueAckCh <- MqttQueueResult{0, err}
+	shouldLoop := true
+	for shouldLoop {
+		// We read two channels for writing out packets. The directSend
+		// channel received synchronous, which has an unbuffered queue.
+		// The queuedSend can back up because the client already has an ID to
+		// check for the response.
+		// NOTE: Writes are sent on a bufio.Writer, so writes are almost
+		// guaranteed to succeed, even if they never reach the server. See the
+		// note in writePacket for more information.
+		select {
+		case pktSendData := <-conn.directSendPktCh:
+			resultCh := pktSendData.resultCh
+			// Verify that we have the write packet type for direct sends
+			if pktSendData.pkt == nil {
+				// Case where the channel is closed
+				shouldLoop = false
+				break
+			}
+			if conn.verifyPacketType(pktSendData.pkt.Type()) != directPacketSendType {
+				resultCh <- MqttResponse{
+					Err: errors.New("Packet type must be queued"),
+				}
+			} else {
```

Comment:
In Go lang, `else` statement is omitted if it returns by kind of err. This case is the following pattern, I think (It is not so if it may is an another something reasons).

https://golang.org/doc/effective_go.html#control-structures

So...
```go
			if conn.verifyPacketType(pktSendData.pkt.Type()) != directPacketSendType {
				resultCh <- MqttResponse{
					Err: errors.New("Packet type must be queued"),
				}
				continue
			}
			resultCh <- MqttResponse{Err: conn.writePacket(pktSendData.pkt)}
```

Reply:
I removed the "continue" so the code is more symmetric, so now the "else" is necessary.